File: tests/js/test_xem_html_trong_javis.js
/* Mở landing/HTML trên Javis phải còn nút Về Javis — không kẹt trang thuần.

       node tests/js/test_xem_html_trong_javis.js

   Chủ repo 2026-09-13: tạo landing từ skill, bấm mở trên Javis — không có nút thoát
   về Javis / đóng. Gốc: .html mở mã nguồn; "Mở tab mới" đi /files/raw; PWA standalone
   thường thay chính cửa sổ app, landing không có chrome Javis.

   Hợp đồng: xem trang trong iframe, thanh Javis còn, tab mới qua /files/html-view. */
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..", "..");
const D = (f) => fs.readFileSync(path.join(ROOT, "dashboard", f), "utf8");
const CONSOLE = D("console.js");
const FE = D("file-editor.js");
const CR = D("chat-render.js");
const CSS = D("style.css");
const INDEX = D("index.html");

const fails = [];
const check = (name, cond) => { console.log((cond ? "ok   " : "FAIL ") + name); if (!cond) fails.push(name); };

check("console.js có _neRenderHtmlPage", /function _neRenderHtmlPage\s*\(/.test(CONSOLE));
check("mặc định xem trang (iframe /files/raw)",
  /ne-body ne-html mode-preview/.test(CONSOLE) && /iframe class="ne-frame"/.test(CONSOLE));
check("có nút gat Xem trang / Sửa mã",
  /Xem trang/.test(CONSOLE) && /Sửa mã/.test(CONSOLE));
check("nút đóng HTML ghi Về Javis",
  /← Về Javis/.test(CONSOLE) && /ne-back/.test(CONSOLE));
check("Mở tab mới HTML đi html-view, không raw",
  /_moFileNgoai/.test(CONSOLE)
  && /_vtHtmlView/.test(CONSOLE)
  && /\/files\/html-view\?/.test(CONSOLE)
  && /window\.open\(_vtHtmlView\(rel\)/.test(CONSOLE));
check("CANARY: PWA không window.open raw HTML",
  /_laPwa/.test(CONSOLE) && /display-mode:\s*standalone/.test(CONSOLE)
  && CONSOLE.indexOf("_laPwa()") < CONSOLE.indexOf("window.open(_vtHtmlView"));
check("openNote .html đi _neRenderHtmlPage, vẫn ghim",
  /ext === "\.html"/.test(CONSOLE) && /_neRenderHtmlPage\(/.test(CONSOLE)
  && /JavisPin\.set\(\{ name: d\.name \|\| it\.name \|\| rel, rel, abs: d\.abs \}\)/.test(CONSOLE));
check("_neCommonBtns gọi _moFileNgoai (HTML không nhảy raw trực tiếp)",
  /_moFileNgoai\(rel\)/.test(CONSOLE));

check("file-editor có renderHtml", /function renderHtml\s*\(/.test(FE));
check("file-editor HTML mặc định iframe + Về Javis",
  /jvfe-html/.test(FE) && /← Về Javis/.test(FE) && /function htmlViewUrl/.test(FE));
check("file-editor PWA chặn nhảy raw",
  /laPwa\(\)/.test(FE) && /e\.preventDefault\(\)/.test(FE));
check("CANARY: không lưu HTML khi mã nguồn chưa tải (tránh ghi đè file rỗng)",
  /ready\(\)|typeof ready === "function"/.test(FE) && /return loaded/.test(FE));
check("file-editor vẫn có openLink (test_code_hl)",
  /function openLink\(/.test(FE) && /JavisCodeHL\.attach/.test(FE));

check("chat-render: link .html title Xem trang",
  /Xem trang/.test(CR) && /\.html\?\$/.test(CR));
check("chat-render: .md vẫn Mở ra sửa", /Mở ra sửa/.test(CR));

check("CSS: ne-html preview ẩn mã nguồn, source ẩn iframe",
  /\.ne-body\.ne-html\.mode-preview \.ne-src/.test(CSS)
  && /\.ne-body\.ne-html\.mode-source \.ne-frame/.test(CSS)
  && /\.ne-actions button\.ne-back/.test(CSS));

const cacheV = (f) => Number((INDEX.match(new RegExp(f.replace(/\./g, "\\.") + "\\?v=(\\d+)")) || [])[1] || 0);
check("index.html cache-bust console / file-editor / chat-render / style",
  cacheV("console.js") >= 149
  && cacheV("file-editor.js") >= 9
  && cacheV("chat-render.js") >= 15
  && cacheV("style.css") >= 93);

console.log();
if (fails.length) { console.log(fails.length + " test HỎNG: " + fails.join(", ")); process.exit(1); }
console.log("OK - test_xem_html_trong_javis: tất cả pass");

File: tests/python/test_xem_html_trong_javis.py
"""Landing/HTML mở trên Javis phải còn nút Về Javis — không kẹt trang thuần.

    python tests/python/test_xem_html_trong_javis.py

Không mạng. Dùng JAVIS_STATE_DIR tạm.
"""
import os
import sys
import tempfile
from pathlib import Path

from _paths import ROOT, SERVER  # noqa: E402,F401

try:
    sys.stdout.reconfigure(encoding="utf-8")
except Exception:
    pass

os.environ.setdefault("JAVIS_STATE_DIR", tempfile.mkdtemp(prefix="javis-htmlview-"))

import main  # noqa: E402

fails = []


def check(name: str, cond: bool) -> None:
    print(("ok   " if cond else "FAIL ") + name)
    if not cond:
        fails.append(name)


_TMP = Path(tempfile.mkdtemp(prefix="javis-htmlview-brain-")).resolve()
BRAIN = _TMP / "brains" / "Bo Nao"
land = BRAIN / "exports" / "landing" / "demo"
land.mkdir(parents=True)
(land / "index.html").write_text(
    "<!doctype html><html><body><h1>Landing demo</h1></body></html>",
    encoding="utf-8",
)
(BRAIN / "ghi-chu.md").write_text("# note", encoding="utf-8")

main._brain_root = lambda brain: str(BRAIN)
main.BRAINS_DIR = str(_TMP / "brains")
os.environ["JAVIS_HOST"] = "0.0.0.0"
os.environ.pop("JAVIS_FILES_ROOT", None)
os.environ.pop("JAVIS_REQUIRE_LOGIN", None)

rel = "exports/landing/demo/index.html"

wrap = main.html_view_response("brain", rel)
body = wrap.body.decode("utf-8") if hasattr(wrap, "body") else ""
check("html-view trả HTMLResponse", getattr(wrap, "status_code", 200) == 200)
check("html-view có nút Về Javis về /", "← Về Javis" in body and 'href="/"' in body)
check("html-view nhúng iframe raw (plain=1), không dán nội dung landing vào khung",
      "iframe" in body and "plain=1" in body and "Landing demo" not in body)
check("html-view không sửa file xuất — chỉ bọc",
      "<h1>Landing demo</h1>" not in body)

bad = main.html_view_response("brain", "../etc/passwd")
check("html-view chặn traversal",
      getattr(bad, "status_code", 0) in (400, 404)
      or (hasattr(bad, "body") and b"error" in bad.body))

md = main.html_view_response("brain", "ghi-chu.md")
check("html-view từ chối file không phải HTML", getattr(md, "status_code", 0) == 400)

redir = main.raw_file_response("brain", rel, dest="document")
loc = ""
if hasattr(redir, "headers"):
    loc = redir.headers.get("location") or redir.headers.get("Location") or ""
check("raw HTML mở thành tab (dest=document) → html-view",
      getattr(redir, "status_code", 0) in (302, 307) and "html-view" in loc)

frame = main.raw_file_response("brain", rel, dest="iframe")
check("iframe vẫn nhận file thô (landing chạy trong trình sửa)",
      getattr(frame, "status_code", 200) == 200
      and "FileResponse" in type(frame).__name__)

plain = main.raw_file_response("brain", rel, dest="document", plain=True)
check("plain=1 giữ trang thuần, không vòng redirect",
      getattr(plain, "status_code", 200) == 200
      and "FileResponse" in type(plain).__name__)

note = main.raw_file_response("brain", "ghi-chu.md", dest="document")
check("raw .md không bị đẩy sang html-view",
      "html-view" not in str(getattr(note, "headers", {}))
      and "FileResponse" in type(note).__name__)

src = (ROOT / "server" / "main.py").read_text(encoding="utf-8")
check("có route /files/html-view", '@app.get("/files/html-view")' in src)
check("files_raw đọc Sec-Fetch-Dest, không gọi handler khác như hàm",
      "sec-fetch-dest" in src and "def html_view_response(" in src)
check("brain_file_compat không bắt buộc Request (test cũ gọi thẳng handler)",
      "request: Request = None" in src)

print()
if fails:
    print(f"FAIL - test_xem_html_trong_javis: {len(fails)} lỗi: " + ", ".join(fails))
    raise SystemExit(1)
print("OK - test_xem_html_trong_javis: tất cả pass")
